File: typescript/src/number_node.ts
import {
  ArithmeticResultNodeWrapper,
  ArithmeticVariableNode,
  NumberNode,
} from './instruction_serialiser/instruction_serialiser_pb';
import { CalculatableNode } from './calculatable_node';

declare module './instruction_serialiser/instruction_serialiser_pb' {
  interface NumberNode extends CalculatableNode<number> {}
}

NumberNode.prototype.calculate = function(): number {
  return this.getValue();
};

function createNumberNode(num: number): ArithmeticResultNodeWrapper {
  return new ArithmeticResultNodeWrapper().setNumberNode(
    new NumberNode().setValue(num)
  );
}

export { NumberNode, createNumberNode };

File: rust/src/number_node.rs
use std::collections::HashMap;
use crate::arithmetic_type::ArithmeticType;
use crate::calculation_error::CalculationError;
use crate::generic_node::GenericNode;
use crate::instruction_serialiser::ArithmeticResultNodeWrapper;
use crate::instruction_serialiser::arithmetic_result_node_wrapper::Node;
use crate::parameter::Parameter;

pub use crate::instruction_serialiser::NumberNode;

impl GenericNode<ArithmeticType> for NumberNode {
    fn calculate(&self, _: &HashMap<&str, Parameter>) -> Result<ArithmeticType, CalculationError> {
        Ok(self.value)
    }
}

impl NumberNode {
    /// Creates a new ArithmeticResultNodeWrapper wrapping a number node with the value passed to
    /// this function.
    pub fn new(value: ArithmeticType) -> ArithmeticResultNodeWrapper {
        ArithmeticResultNodeWrapper{
            node: Some(
                Node::NumberNode(
                    NumberNode { value }
                )
            )
        }
    }
}